cpp/daal/src/algorithms/kmeans/oneapi/kmeans_dense_lloyd_batch_kernel_ucapi_impl.i

```diff
@@ -94,186 +107,127 @@ Status KMeansDenseLloydBatchKernelUCAPI<algorithmFPType>::compute(const NumericT
 
     Status st;
 
-    auto & context        = Environment::getInstance()->getDefaultExecutionContext();
-    auto & kernel_factory = context.getClKernelFactory();
-
     NumericTable * ntData         = const_cast<NumericTable *>(a[0]);
     NumericTable * ntInCentroids  = const_cast<NumericTable *>(a[1]);
     NumericTable * ntOutCentroids = const_cast<NumericTable *>(r[0]);
     NumericTable * ntAssignments  = const_cast<NumericTable *>(r[1]);
     NumericTable * ntObjFunction  = const_cast<NumericTable *>(r[2]);
     NumericTable * ntNIterations  = const_cast<NumericTable *>(r[3]);
 
-    const size_t nIter     = par->maxIterations;
-    const size_t nRows     = ntData->getNumberOfRows();
-    const size_t nFeatures = ntData->getNumberOfColumns();
-    const size_t nClusters = par->nClusters;
-
-    auto fptype_name   = oneapi::internal::getKeyFPType<algorithmFPType>();
-    auto build_options = fptype_name;
-
-    build_options.add(getBuildOptions(nClusters));
-
-    services::String cachekey("__daal_algorithms_kmeans_lloyd_dense_batch_");
-    cachekey.add(fptype_name);
-    cachekey.add(build_options.c_str());
-
-    {
-        DAAL_ITTNOTIFY_SCOPED_TASK(compute.buildProgram);
-        kernel_factory.build(ExecutionTargetIds::device, cachekey.c_str(), kmeans_cl_kernels, build_options.c_str());
-    }
-
-    const size_t nPartialCentroids = 128;
-    const size_t nValuesInBlock    = 1024 * 1024 * 1024 / sizeof(algorithmFPType);
-    const size_t nMinRows          = 1;
-    size_t gemmBlockSize           = nValuesInBlock;
-
-    while (gemmBlockSize > nValuesInBlock / nClusters)
-    {
-        gemmBlockSize >>= 1;
-    }
-
-    if (gemmBlockSize < nMinRows)
-    {
-        return Status(ErrorKMeansNumberOfClustersIsTooLarge);
-    }
-
-    size_t datasetBlockSize = nValuesInBlock;
-    while (datasetBlockSize > nValuesInBlock / nFeatures)
+    if (ntData->getNumberOfRows() > static_cast<size_t>(UINT_MAX) || ntData->getNumberOfColumns() > static_cast<size_t>(UINT_MAX)
```

Comment:
Don't we have a special "overflow" check for that case? If not, it'd better create such a define.

Reply:
Last time I was recommended to do it this way